lib/utils/configReader.js

```diff
@@ -1,85 +1,86 @@
 'use strict'
 
 var path = require('path')
-
-var extend = require('lodash/object/extend')
+var warn = require('debug')('risingstack/trace:warn')
 var defaults = require('lodash/object/defaults')
 
-var allowedReporters = ['logstash', 'trace']
-
-function isReporterValid (reporterType) {
-  var index = allowedReporters
-    .indexOf(reporterType)
+function ConfigReader (config) {
+  this.parameterConfig = config || { }
+}
 
-  return index > -1
+ConfigReader.prototype._getEnvVarConfig = function () {
+  return {
+    collectInterval: process.env.TRACE_COLLECT_INTERVAL,
+    initialSampleRate: process.env.TRACE_INITIAL_SAMPLE_RATE,
+    collectorApi: process.env.TRACE_COLLECTOR_API_URL,
+    appName: process.env.TRACE_APP_NAME,
+    configPath: process.env.TRACE_CONFIG_PATH,
+    apiKey: process.env.TRACE_API_KEY
+  }
 }
 
-function ConfigReader (collectorConfig) {
-  this.collectorConfig = collectorConfig
+ConfigReader.prototype._getDefaultConfig = function () {
+  return require('../config')
 }
 
-ConfigReader.prototype._require = function (filePath) {
-  return require(filePath)
+ConfigReader.prototype._readConfigFile = function (file) {
+  return require(file)
 }
 
-ConfigReader.prototype._initConfig = function () {
-  var config
-  try {
-    var configPath = this.collectorConfig.configPath ||
-      path.join(__dirname, '../../../../../', 'trace.config.js')
-    var configToExtend = this._require(configPath)
-    config = extend({}, configToExtend)
-  } catch (ex) {
-    throw new Error('Invalid trace.config.js configuration file')
+ConfigReader.prototype._getFileConfig = function (file) {
+  if (file) {
+    try {
+      return this._readConfigFile(path.resolve(file))
+    } catch (ex) {
+      if (ex.code === 'MODULE_NOT_FOUND') {
+        warn('Configuration file not found')
+        return { }
+      } else {
+        throw new Error('Invalid trace.config.js configuration file')
+      }
+    }
+  } else /* no file path given */ {
+    return { }
   }
+}
 
-  return config
+ConfigReader.prototype._getReporterConfig = function (reporterArgs) {
+  return this._require('../reporters').trace.create(reporterArgs)
 }
 
-ConfigReader.prototype._getReporterConfig = function () {
-  var reporterConfig
+ConfigReader.prototype.getConfig = function () {
+  var parameterConfig = this.parameterConfig
+  var envVarConfig = this._getEnvVarConfig()
+  var defaultConfig = this._getDefaultConfig()
 
-  if (!isReporterValid(this.collectorConfig.reporterType)) {
-    throw new Error('Invalid reporter type')
-  }
+  var config = {}
 
-  if (!this.collectorConfig.reporterConfig) {
-    throw new Error('No reporter config')
-  }
+  defaults(config, parameterConfig, envVarConfig)
```

Comment:
the resulting object is mutatated twice later, consistent renaming would result in bloated partial config names. I'd rather do this with implicit mutations.